src/main/java/com/es/phoneshop/web/ProductListPageServlet.java

```diff
@@ -30,6 +39,46 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
                 SortField.getValue(sortField),
                 SortOrder.getValue(sortOrder)
         ));
-        request.getRequestDispatcher("/WEB-INF/pages/productList.jsp").forward(request, response);
+        request.getRequestDispatcher(PRODUCT_LIST_JSP).forward(request, response);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        String productIdString = request.getParameter("id");
+        String quantityString = request.getParameter("quantity");
+
+        Cart cart = cartService.getCart(request);
+        int quantity;
+        try {
+            Long productId = Long.valueOf(productIdString);
+            NumberFormat format = NumberFormat.getInstance(request.getLocale());
+            quantity = format.parse(quantityString).intValue();
+            cartService.add(cart, productId, quantity);
+        } catch (ParseException | NumberFormatException ex) {
+            request.setAttribute("error", "Not a number");
+            doGet(request, response);
+            return;
+        } catch (OutOfStockException e) {
+            request.setAttribute("error", "Out of stock, available " + e.getStockAvailable());
+            doGet(request, response);
+            return;
+        }
+
+        response.sendRedirect(makeRedirectPath(request));
+    }
+
+    private String makeRedirectPath(HttpServletRequest request) {
+        String redirectPath = request.getContextPath() + "/products?";
+        if (request.getParameter("query") != null) {
```

Comment:
It's good idea to make String constant like QUERY_PARAM = "query" cause it's used multiple times. It will help to avoid some missclicks or other errors, just hint, no changes required.